File: 5TH SEMESTER/LB-I/FCFS.java
import java.util.*;

public class FCFS {

  public static void main(String args[]) {
    Scanner sc = new Scanner(System.in);

    System.out.println("Enter No. of Processes: ");

    int n = sc.nextInt();

    int pno[] = new int[n];

    int ar[] = new int[n];

    int bt[] = new int[n];

    int ct[] = new int[n];

    int ta[] = new int[n];

    int wt[] = new int[n];

    int temp;

    float avgwt = 0;

    for (int i = 0; i < n; i++) {
      System.out.println("Enter Process " + (i + 1) + " Arrival Time: ");

      ar[i] = sc.nextInt();

      System.out.println("Enter Process " + (i + 1) + " Burst Time: ");

      bt[i] = sc.nextInt();

      pno[i] = i + 1;
    }

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n - (i + 1); j++) {
        if (ar[j] > ar[j + 1]) {
          temp = ar[j];

          ar[j] = ar[j + 1];

          ar[j + 1] = temp;

          temp = bt[j];

          bt[j] = bt[j + 1];

          bt[j + 1] = temp;

          temp = pno[j];

          pno[j] = pno[j + 1];

          pno[j + 1] = temp;
        }
      }
    }

    for (int i = 0; i < n; i++) {
      if (i == 0) {
        ct[i] = ar[i] + bt[i];
      } else {
        if (ar[i] > ct[i - 1]) {
          ct[i] = ar[i] + bt[i];
        } else ct[i] = ct[i - 1] + bt[i];
      }

      ta[i] = ct[i] - ar[i];

      wt[i] = ta[i] - bt[i];

      avgwt += wt[i];
    }

    System.out.println(
      "\n P.No. \t Arrival Time \t Burst time \t Service time \t Waiting time"
    );

    for (int i = 0; i < n; i++) {
      System.out.println(
        "P" +
        pno[i] +
        " \t " +
        ar[i] +
        "\t\t" +
        bt[i] +
        "\t\t" +
        ct[i] +
        "\t\t" +
        wt[i]
      );
    }

    sc.close();

    System.out.println("");

    System.out.println("Gantt Chart : ");

    for (int i = 0; i < n; i++) {
      int c = 0;

      for (int j = 0; j < ct[i]; j++) {
        if (c == 0) {
          System.out.print("|____P" + pno[i] + " _");

          c = 1;
        } else {
          System.out.print("_");
        }
      }

      System.out.print(" |");
    }

    System.out.println("");

    for (int i = 0; i < n; i++) {
      int c = 0;

      for (int j = 0; j < ct[i]; j++) {
        if (c == 0) {
          System.out.print("\t" + ct[i]);

          c = 1;
        } else {
          System.out.print(" ");
        }
      }

      System.out.print(" |");
    }

    System.out.println("\n\n Average Waiting Time : " + (avgwt / n));
  }
}

File: New folder/Songs/Scheduling/FCFS.cpp
//FCFS:
#include<iostream>
#include<algorithm>
using namespace std;

struct Process {
    int id, bt, at, wt, tat, ct, start_time;
};

bool compareArrivalTime(Process p1, Process p2) {
    return p1.at < p2.at;
}

int main() {
    int n;
    float avg_tat = 0, avg_wt = 0;
    cout<<"Enter no. : ";
    cin >> n;
    Process p[n];

    for (int i = 0; i < n; i++) {
        p[i].id = i + 1;
        cout<<"Enter bt for process "<<i+1<<" : ";
        cin >> p[i].bt;
        cout<<"Enter at for process "<<i+1<<" : ";
        cin >> p[i].at;
    }

    sort(p, p + n, compareArrivalTime);

    p[0].start_time = p[0].at;
    p[0].ct = p[0].start_time + p[0].bt;
    p[0].tat = p[0].ct - p[0].at;
    p[0].wt = p[0].tat - p[0].bt;

    for (int i = 1; i < n; i++) {
        p[i].start_time = max(p[i - 1].ct, p[i].at);
        p[i].ct = p[i].start_time + p[i].bt;
        p[i].tat = p[i].ct - p[i].at;
        p[i].wt = p[i].tat - p[i].bt;
    }

    cout << "\nProcess ID | Arrival Time | Burst Time | Turnaround Time | Waiting Time\n";
    for (int i = 0; i < n; i++) {
        cout << "P" << p[i].id << "    | " << p[i].at << "         | " << p[i].bt
             << "        | " << p[i].tat << "           | " << p[i].wt << endl;
    }

    cout << "\nGantt chart:\n";
    for (int i = 0; i < n; i++) cout << "p" << p[i].id << " |";
    cout << endl;
    for (int i = 0; i < n; i++) cout << p[i].start_time << "\t";
    cout << p[n-1].ct << endl;

    for (int i = 0; i < n; i++) {
        avg_wt += p[i].wt;
        avg_tat += p[i].tat;
    }

    cout << "\nAverage turnaround time is: " << avg_tat / n;
    cout << "\nAverage waiting time is: " << avg_wt / n;

    return 0;
}

